### PR TITLE
Rewrite code for mapping source positions to HTML ASTs

### PR DESCRIPTION
**Pull request by rictic**
This fixes a bunch of bugs, including Polymer/atom-plugin#7 Polymer/atom-plugin#8 and Polymer/atom-plugin#9 as well as a *lot* of subtle edge cases that we weren't getting right.

There's probably more work to be done here, I didn't test any multiline use cases here, though I'm hoping that since this builds on #347 that most of that stuff should Just Work.

Also I made some methods protected that should have been (they were only ever intended to be defined or called by subclasses anyways).

/cc @TimvdLippe

### CODE REVIEW

**Comment by TimvdLippe** [comment]
Confirmed it fixes most of the issues :tada: 

Some issues I still observed:

1. Type `<dom-module></dom-module>`
* Type `id=""` in the start tag
* Type `be` inside the quotes and observe that the suggestions list shows `<behavior-test-elem>`. Note that actually letting it autocomplete does not work (known issue).

And another one:

1. Type `<beh` and autocomplete to `<behavior-test-elem>`
* Go back to the start tag and remove the space
* Hit ctrl + space before the `>` of the start tag
* Observe that no suggestions are shown

**Comment by rictic** [comment]
Hm. So for the first issue, I can reproduce something similar. I'm seeing autocompletions of attribute names for attribute values, which I tried but apparently did not succeed in fixing in this PR. Will take another shot at that.

 I haven't been able to reproduce getting element name suggestions for attribute values though. Mind pasting in the exact text of the file that you're working with, with a `|` where the cursor is?


For the second issue  `<behavior-test-elem|>`, is https://github.com/Polymer/atom-plugin/issues/12 a fair summary of the ideal behavior?

**Comment by TimvdLippe** [comment]
For the second issue instead of showing docs, I would prefer to have attribute completion with the space added in front. The heuristic would probably that if the start tag is closed with a > then it should not autocomplete on the element name but the attributes instead.

**Comment by rictic** [comment]
The logic that I'm going for here is that you get suggestions based on what would happen if you started typing at the current cursor position.

I can see a user wanting either docs for that tag and I can see them wanting to get attribute completions. Both seem like reasonable expectations to me, and I lean towards keeping it as is (but with displaying docs) for consistency reasons.

File: src/editor-service/ast-from-source-position.ts [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+import {ParsedHtmlDocument} from '../html/html-document';
+import {SourceRange} from '../model/model';
+
+import {SourcePosition} from './editor-service';
+
+
+export type LocationResult = LocatedAttribute | LocatedAttributeValue |
+    LocatedTag | LocatedEndTag | LocatedInText | LocatedInScript |
+    LocatedInStyle | LocatedInComment;
+/** In the tagname of a start tag. */
```
Comment (justinfagnani) [inline src/editor-service/ast-from-source-position.ts]: nit: add some newlines

File: src/editor-service/ast-from-source-position.ts [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+}
+/** In an end tag. */
+export interface LocatedEndTag {
+  kind: 'endTag';
+  element: parse5.ASTNode;
+}
+/** In the attributes section of a start tag. Maybe in an attribute name. */
+export interface LocatedAttribute {
+  kind: 'attribute';
+  attribute: string|null;
```
Comment (justinfagnani) [inline src/editor-service/ast-from-source-position.ts]: what does the value `null` mean?

File: src/editor-service/ast-from-source-position.ts [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+import {SourcePosition} from './editor-service';
+
+
+export type LocationResult = LocatedAttribute | LocatedAttributeValue |
+    LocatedTag | LocatedEndTag | LocatedInText | LocatedInScript |
+    LocatedInStyle | LocatedInComment;
+/** In the tagname of a start tag. */
+export interface LocatedTag {
+  kind: 'tagName';
+  element: parse5.ASTNode;
```
Comment (justinfagnani) [inline src/editor-service/ast-from-source-position.ts]: consider naming this `node` so it's consistent across all `LocationResults`

File: src/editor-service/ast-from-source-position.ts [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+/**
+ * Given a position and an HTML document, try to describe what new text typed
+ * at the given position would be.
+ *
+ * Where possible we try to return the ASTNode describing that position, but
+ * sometimes there does not actually exist one. (for a simple case, the empty
+ * string should be interpreted as a text node, but there is no text node in
+ * an empty document, but there would be after the first character was typed).
+ */
+export function getLocationInfoForPosition(
```
Comment (justinfagnani) [inline src/editor-service/ast-from-source-position.ts]: future: definitely seems like this could be a method on Document

File: src/editor-service/ast-from-source-position.ts [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+    return {kind: 'text'};
+  }
+  return location;
+}
+function _getLocationInfoForPosition(
+    node: parse5.ASTNode, position: SourcePosition,
+    document: ParsedHtmlDocument): undefined|LocationResult {
+  const sourceRange = document.sourceRangeForNode(node);
+  const location = node.__location;
+  if (!(sourceRange && location)) {
```
Comment (justinfagnani) [inline src/editor-service/ast-from-source-position.ts]: can you document the reasoning behind this condition? it seems to be that if you can't get a SourceRange for a node, you recurse into its children, but how would you know that's the right thing to do if you can't get a SourceRange? Naively I would think this case should return `null`.

File: src/editor-service/ast-from-source-position.ts [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+  }
+
+  const locationInChildren = _findLocationInChildren(node, position, document);
+  if (locationInChildren) {
+    return locationInChildren;
+  }
+
+  /**
+   * TODO(rictic): upstream the fact that regular locations can have attrs
+   * sometimes.
```
Comment (justinfagnani) [inline src/editor-service/ast-from-source-position.ts]: what does that mean?

File: src/editor-service/ast-from-source-position.ts [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+   */
+  const attrs: parse5.AttributesLocationInfo =
+      (isElementLocationInfo(location) && location.startTag.attrs) ||
+      location['attrs'] || {};
+
+  for (const attrName in attrs) {
```
Comment (justinfagnani) [inline src/editor-service/ast-from-source-position.ts]: consider factoring this into a function

File: src/editor-service/ast-from-source-position.ts [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+
+  const startTagRange = document.sourceRangeForStartTag(node);
+  const endTagRange = document.sourceRangeForEndTag(node);
+
+  // If we're in the end tag... we're in the end tag.
+  if (isPositionInsideRange(position, endTagRange, false)) {
+    return {kind: 'endTag', element: node};
+  }
+
+  if (startTagRange && isPositionInsideRange(position, startTagRange, false)) {
```
Comment (justinfagnani) [inline src/editor-service/ast-from-source-position.ts]: why a truthy check on `startTagRange` here, but not on `endTagRange` above?

File: src/editor-service/ast-from-source-position.ts (hunk 6 above)
Comment (rictic) [inline src/editor-service/ast-from-source-position.ts]: It's a typings thing, expanded the comment.

File: src/editor-service/ast-from-source-position.ts (hunk 7 above)
Comment (rictic) [inline src/editor-service/ast-from-source-position.ts]: Done

File: src/editor-service/ast-from-source-position.ts (hunk 8 above)
Comment (rictic) [inline src/editor-service/ast-from-source-position.ts]: For the compiler's sake, because we use startTagRange inside the if.
